components/layout_2020/replaced.rs

```diff
@@ -276,9 +280,67 @@ impl ReplacedContent {
     pub fn make_fragments(
         &self,
         style: &ServoArc<ComputedValues>,
+        containing_block: &ContainingBlock,
         size: PhysicalSize<Au>,
     ) -> Vec<Fragment> {
-        let rect = PhysicalRect::new(PhysicalPoint::origin(), size);
+        let aspect_ratio = self.preferred_aspect_ratio(&containing_block.into(), style);
+        let natural_size = PhysicalSize::new(
+            self.natural_size.width.unwrap_or(size.width),
+            self.natural_size.height.unwrap_or(size.height),
+        );
+
+        let object_fit_size = aspect_ratio.map_or(size, |aspect_ratio| {
+            let preserve_aspect_ratio_with_comparison =
+                |size: PhysicalSize<Au>, comparison: fn(&Au, &Au) -> bool| {
+                    let (width_axis, height_axis) = if style.writing_mode.is_horizontal() {
+                        (Direction::Inline, Direction::Block)
+                    } else {
+                        (Direction::Block, Direction::Inline)
+                    };
+
+                    let candidate_size = PhysicalSize::new(
+                        aspect_ratio.compute_dependent_size(width_axis, size.height),
+                        size.height,
+                    );
+
+                    if comparison(&candidate_size.width, &size.width) {
+                        return candidate_size;
+                    }
+
+                    debug_assert!(comparison(&candidate_size.height, &size.height));
```

Comment:
Here we know `candidate_size.height == size.height`. The useful assert is with the new candidate height, `aspect_ratio.compute_dependent_size(height_axis, size.width)`.

It's also slightly shorter if the candidate width/height is stored in a variable and the `PhysicalSize` is created later, but fine either way.

Reply:
I didn't realize you flipped the order of the axes in your suggestion. This should be fixed. now.